lib/nx_signal/filters.ex

```diff
@@ -25,4 +25,57 @@ defmodule NxSignal.Filters do
     # Define sinc(0) = 1
     Nx.select(zero_idx, 1, Nx.sin(t) / t)
   end
+
+  defn median_filter(t, opts \\ [kernel_size: 3]) do
+    {height, width} = Nx.shape(t)
+    kernel_tensor = Nx.broadcast(0.0, {opts[:kernel_size]})
+    output = Nx.broadcast(0.0, Nx.shape(t))
+
+    {_, _, _, _, _, result} =
+      while {i = 0, t, kernel_tensor, height, width, output}, i < height do
```

Comment:
```suggestion
    {result, _} =
      while {output, {i = 0, t, kernel_tensor, height, width}}, i < height do
```

Refactor the loops to follow this kind of pattern as it's easier to match on and brings the output to the front

Reply:
Actually, here's an idea for a refactor using vectorization.

Let's use hardcoded values for simplicity:

window_shape = {3, 3}
tensor = Nx.iota({5, 5})

```elixir
# first, let's build out a base indices tensor. these can be used with gather to access each element in the tensor
idx = Nx.stack([Nx.iota(tensor.shape, axis: 0), Nx.iota(tensor.shape, axis: 1)], axis: -1) |> Nx.reshape({:auto, 2}) |> Nx.vectorize(:elements)

# then, we can build an offset mask for each window:

padding_x = div(elem(window_shape, 0), 2)
padding_y = div(elem(window_shape, 1), 2)
window_offsets = Nx.stack([
  Nx.iota(window_shape, axis: 0) |> Nx.subtract(padding_x),
  Nx.iota(window_shape, axis: 1)  |> Nx.subtract(padding_y)
], axis: -1) |> Nx.reshape({:auto, 2}) |> Nx.vectorize(:offsets)

# because we're vectorizing with different names, when we operate on the tensors, we'll get
# a cross-operation between each element of one with all of the others. This can be used
# to build each median window.

idx = Nx.add(idx, window_offsets) |> Nx.add(Nx.tensor([padding_x, padding_y]))

tensor
|> Nx.pad(0, [{padding_x, padding_x, 0}, {padding_y, padding_y, 0}])
|> Nx.gather(idx) # the tensor here is now vectorized {elements: 25, offsets: 9} with scalar elements
|> Nx.devectorize(keep_names: false)
|> Nx.median(axis: 1)
|> Nx.reshape(tensor.shape)
```

I haven't cross-checked results, this might be wrong

Reply:
@santiago-imelio did you get around to trying the vectorized version out?

Reply:
Instead of `Nx.pad` the boundaries can be handled using different functions (e.g.`Nx.reflect` or custom implementations) to handle `mode`. I wouldn't worry about this in this PR, though.

Reply:
I haven't yet, first I would like to understand the concept of vectorization and play with it, need some time for that. Also I'm somewhat of a noob to scientific programming so bare with me haha.

Reply:
No worries. I think there is a guide on vectorization in the main nx repo

Reply:
@polvalente I was able to use vectorization, although it's somwhat different from your suggestion. Seems to be working fine with the same examples (with minor discrepancies). Let me know if the approach is correct.